rust/kernel/task.rs

```diff
@@ -0,0 +1,188 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Tasks (threads and processes).
+//!
+//! C header: [`include/linux/sched.h`](../../../../include/linux/sched.h).
+
+use crate::{bindings, c_types};
+use core::{cell::Cell, marker::PhantomData, mem::ManuallyDrop, ops::Deref};
+
+extern "C" {
+    #[allow(improper_ctypes)]
+    fn rust_helper_signal_pending(t: *const bindings::task_struct) -> c_types::c_int;
+    #[allow(improper_ctypes)]
+    fn rust_helper_current() -> *mut bindings::task_struct;
+    #[allow(improper_ctypes)]
+    fn rust_helper_get_task_struct(t: *mut bindings::task_struct);
+    #[allow(improper_ctypes)]
+    fn rust_helper_put_task_struct(t: *mut bindings::task_struct);
+}
+
+/// Wraps the kernel's `struct task_struct`.
+///
+/// # Invariants
+///
+/// The pointer [`Task::ptr`] is non-null and valid. Its reference count is also non-zero.
+///
+/// # Examples
+///
+/// The following is an example of getting the pid of the current thread with zero additional cost
+/// when compared to the C version:
+///
+/// ```
+/// # use kernel::prelude::*;
+/// use kernel::task::Task;
+///
+/// # fn test() {
+/// Task::current().pid();
+/// # }
+/// ```
+///
+/// Getting the pid of the current process, also zero additional cost:
```

Comment:
Should we use the uppercase, i.e. `PID`, as well as for `OK` etc.? Not sure what is most common (you use `PIDs` below too)

Reply:
I was in search of a type that was an acronym in the standard library, the first one I found was `TcpStream` (in golang it would have been TCPStream IIRC), so I followed that.

I have no preference though. Do you think `PID` would be better?

Reply:
Ah, I meant for the comments, not the type names.

Reply:
Oh, I see. Yes, I'll change them.

Reply:
`getpid()` manpage and Wikipedia use the uppercase. POSIX uses "process ID". So I guess going with uppercase (for the prose) is better.